Sources/AsyncHTTPClient/ConnectionPool.swift

```diff
@@ -0,0 +1,509 @@
+    private var _connectionProviders: [Key: ConnectionProvider] = [:]
+
+    private let lock = Lock()
+
+    init(group: EventLoopGroup, configuration: HTTPClient.Configuration) {
+        self.loopGroup = group
+        self.configuration = configuration
+    }
+
+    /// This enables the connection pool to store and retreive `ConnectionProvider`s
+    /// by ensuring thread safety using the `lock`.
+    private subscript(key: Key) -> ConnectionProvider? {
+        get {
+            return self.lock.withLock {
+                _connectionProviders[key]
+            }
+        }
+        set {
+            self.lock.withLock {
+                _connectionProviders[key] = newValue
+            }
+        }
+    }
+
+    /// This method asks the pool for a connection usable by the specified `request`, respecting the specified options.
+    ///
+    /// - parameter request: The request that needs a `Connection`
+    /// - parameter preference: The `EventLoopPreference` the connection pool will respect to lease a new connection
+    /// - parameter deadline: The connection timeout
+    /// - Returns: A connection  corresponding to the specified parameters
+    ///
+    /// When the pool is asked for a new connection, it creates a `Key` from the url associated to the `request`. This key
+    /// is used to determine if there already exists an associated `ConnectionProvider` in `connectionProviders`
+    /// if it does, the connection provider then takes care of leasing a new connection. If a connection provider doesn't exist, it is created.
+    func getConnection(for request: HTTPClient.Request, preference: HTTPClient.EventLoopPreference, deadline: NIODeadline?) -> EventLoopFuture<Connection> {
+        self.lock.withLock {
+            let key: Key
+            // Created a key and checks the URL components are valid
+            do {
+                key = try Key(url: request.url)
+            } catch {
+                return self.loopGroup.next().makeFailedFuture(error)
+            }
+
+            if let provider = self._connectionProviders[key] {
+                // If a connection provider already exists for the given key,
+                // directly ask it for a new connection
+                return provider.getConnection(preference: preference)
+            } else {
+                // If no connection provider exists for the given key, create a new one
+                let selectedEventLoop: EventLoop
+                switch preference.preference {
+                case .indifferent:
+                    selectedEventLoop = self.loopGroup.next()
+                case .prefers(let preferred):
+                    selectedEventLoop = preferred
+                }
+
+                // Completed when we have been able to determine the connection
+                // should be using HTTP/1.1 or HTTP/2.
+                let protocolChoosenPromise = self.loopGroup.next().makePromise(of: ConnectionProvider.self)
```

Comment:
Is there any reason this promise doesn't use the event loop chosen above?

Reply:
No, I missed it thank you! Will fix it :)